File: iosApp/iosApp/Modules/NFTsCollection ✅/NFTsCollectionWireframe.swift
// Created by web3d4v on 29/05/2022.
// Copyright (c) 2022 Sons Of Crypto.
// SPDX-License-Identifier: MIT

import UIKit
import web3walletcore

final class DefaultNFTsCollectionWireframe {
    private weak var parent: UIViewController?
    private let context: NFTsCollectionWireframeContext
    private let nftDetailWireframeFactory: NFTDetailWireframeFactory
    private let nftsService: NFTsService

    private weak var vc: UIViewController?
    
    init(
        _ parent: UIViewController?,
        context: NFTsCollectionWireframeContext,
        nftDetailWireframeFactory: NFTDetailWireframeFactory,
        nftsService: NFTsService
    ) {
        self.parent = parent
        self.context = context
        self.nftDetailWireframeFactory = nftDetailWireframeFactory
        self.nftsService = nftsService
    }
}

extension DefaultNFTsCollectionWireframe: NFTsCollectionWireframe {

    func present() {
        let vc = wireUp()
        parent?.show(vc, sender: self)
    }

    func navigate(destination: NFTsCollectionWireframeDestination) {
        if let input = destination as? NFTsCollectionWireframeDestination.NFTDetail {
            let context = NFTDetailWireframeContext(
                nftId: input.identifier,
                collectionId: context.collectionId
            )
            nftDetailWireframeFactory.make(vc, context: context).present()
        }
        if destination is NFTsCollectionWireframeDestination.Dismiss {
            vc?.popOrDismiss()
        }
    }
}

private extension DefaultNFTsCollectionWireframe {

    func wireUp() -> UIViewController {
        let vc: NFTsCollectionViewController = UIStoryboard(.nftsCollection).instantiate()
        let interactor = DefaultNFTsCollectionInteractor(nftsService: nftsService)
        let presenter = DefaultNFTsCollectionPresenter(
            view: WeakRef(referred: vc),
            wireframe: self,
            interactor: interactor,
            context: context
        )
        vc.presenter = presenter
        self.vc = vc
        guard parent?.asNavVc == nil else { return vc }
        let nc = NavigationController(rootViewController: vc)
        self.vc = nc
        return nc
    }
}

File: androidApp/src/main/java/com/sonsofcrypto/web3wallet/android/modules/compose/nftscollection/NFTsCollectionWireframe.kt
package com.sonsofcrypto.web3wallet.android.modules.compose.nftscollection

import androidx.fragment.app.Fragment
import com.sonsofcrypto.web3lib.utils.WeakRef
import com.sonsofcrypto.web3wallet.android.common.NavigationFragment
import com.sonsofcrypto.web3wallet.android.modules.compose.nftdetail.NFTDetailWireframeFactory
import com.sonsofcrypto.web3walletcore.modules.nftDetail.NFTDetailWireframeContext
import com.sonsofcrypto.web3walletcore.modules.nftsCollection.*
import com.sonsofcrypto.web3walletcore.services.nfts.NFTsService

class DefaultNFTsCollectionWireframe(
    private val parent: WeakRef<Fragment>?,
    private val context: NFTsCollectionWireframeContext,
    private val nftsService: NFTsService,
    private val nftDetailWireframeFactory: NFTDetailWireframeFactory,
): NFTsCollectionWireframe {

    override fun present() {
        val fragment = wireUp()
        (parent?.get() as? NavigationFragment)?.push(fragment, true)
    }

    override fun navigate(destination: NFTsCollectionWireframeDestination) {
        when (destination) {
            is NFTsCollectionWireframeDestination.NFTDetail -> {
                val context = NFTDetailWireframeContext(
                    destination.identifier, context.collectionId
                )
                nftDetailWireframeFactory.make(
                    parent?.get(),
                    context
                ).present()
            }
            is NFTsCollectionWireframeDestination.Dismiss -> {
                (parent?.get() as? NavigationFragment)?.pop()
            }
        }
    }

    private fun wireUp(): Fragment {
        val view = NFTsCollectionFragment()
        val interactor = DefaultNFTsCollectionInteractor(nftsService)
        val presenter = DefaultNFTsCollectionPresenter(
            WeakRef(view),
            this,
            interactor,
            context,
        )
        view.presenter = presenter
        return view
    }
}